File: backend/app/schemas/movement.py
import uuid
from typing import Optional
from datetime import datetime
from pydantic import BaseModel


class GateOTPRequest(BaseModel):
    movement_type: str  # OUT or IN
    destination: Optional[str] = None
    expected_return_iso: Optional[str] = None  # ISO datetime string


class GateOTPResponse(BaseModel):
    otp: str
    expires_in_seconds: int
    student_id: str
    movement_type: str


class GuardConfirmRequest(BaseModel):
    student_id: uuid.UUID
    otp: str
    movement_type: str


class GuardConfirmResponse(BaseModel):
    success: bool
    student_name: str
    room_number: str
    photo_url: Optional[str] = None
    movement_type: str
    timestamp: datetime


class MovementLogResponse(BaseModel):
    id: uuid.UUID
    student_name: Optional[str] = None
    room_number: Optional[str] = None
    type: str
    destination: Optional[str] = None
    expected_return: Optional[datetime] = None
    actual_return: Optional[datetime] = None
    is_overdue: bool
    is_flagged: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class ManualEntryRequest(BaseModel):
    room_number: str
    movement_type: str  # OUT or IN
    note: Optional[str] = None


class LiveStatusResponse(BaseModel):
    students_out: list[MovementLogResponse]
    total_out_count: int
    overdue_count: int

File: mobile/packages/dormconnect_core/lib/src/models/movement.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'movement.freezed.dart';
part 'movement.g.dart';

enum MovementType { out, in_, manual }

@freezed
class MovementLog with _$MovementLog {
  const factory MovementLog({
    required String id,
    required String studentId,
    String? studentName,
    String? roomNumber,
    String? photoUrl,
    required MovementType type,
    String? destination,
    DateTime? expectedReturn,
    DateTime? actualReturn,
    @Default(false) bool isOverdue,
    @Default(false) bool isFlagged,
    String? note,
    DateTime? createdAt,
  }) = _MovementLog;

  factory MovementLog.fromJson(Map<String, dynamic> json) =>
      _$MovementLogFromJson(json);
}

@freezed
class GateOtp with _$GateOtp {
  const factory GateOtp({
    required String otp,
    required int expiresInSeconds,
    required String studentId,
    String? movementType,
  }) = _GateOtp;

  factory GateOtp.fromJson(Map<String, dynamic> json) =>
      _$GateOtpFromJson(json);
}

@freezed
class LiveRequest with _$LiveRequest {
  const factory LiveRequest({
    required String studentId,
    required String otp,
    required String studentName,
    String? roomNumber,
    String? rollNumber,
    String? photoUrl,
    required int secondsRemaining,
  }) = _LiveRequest;

  factory LiveRequest.fromJson(Map<String, dynamic> json) =>
      _$LiveRequestFromJson(json);
}
